File: packages/core/src/version/VersionNegotiator.ts
import { ErrorCode, RynBridgeError } from '../errors.js';

export interface SemVer {
  major: number;
  minor: number;
  patch: number;
}

export class VersionNegotiator {
  static parse(version: string): SemVer {
    const match = version.match(/^(\d+)\.(\d+)\.(\d+)$/);
    if (!match) {
      throw new RynBridgeError(
        ErrorCode.VERSION_MISMATCH,
        `Invalid version format: "${version}". Expected "major.minor.patch"`,
        { version },
      );
    }

    return {
      major: Number(match[1]),
      minor: Number(match[2]),
      patch: Number(match[3]),
    };
  }

  static isCompatible(local: string, remote: string): boolean {
    const localVer = VersionNegotiator.parse(local);
    const remoteVer = VersionNegotiator.parse(remote);

    if (localVer.major === 0 && remoteVer.major === 0) {
      return localVer.minor === remoteVer.minor;
    }

    return localVer.major === remoteVer.major;
  }

  static assertCompatible(local: string, remote: string): void {
    if (!VersionNegotiator.isCompatible(local, remote)) {
      throw new RynBridgeError(
        ErrorCode.VERSION_MISMATCH,
        `Version mismatch: local ${local} is incompatible with remote ${remote}`,
        { local, remote },
      );
    }
  }
}

File: android/core/src/main/kotlin/com/devryner/rynbridge/core/VersionNegotiator.kt
package com.devryner.rynbridge.core

data class SemVer(
    val major: Int,
    val minor: Int,
    val patch: Int
) {
    val description: String get() = "$major.$minor.$patch"
}

class VersionNegotiator {

    fun parse(version: String): SemVer {
        val parts = version.split(".")
        if (parts.size != 3) {
            throw RynBridgeError(
                code = ErrorCode.VERSION_MISMATCH,
                message = "Invalid version format: '$version'"
            )
        }
        val major = parts[0].toIntOrNull()
        val minor = parts[1].toIntOrNull()
        val patch = parts[2].toIntOrNull()
        if (major == null || minor == null || patch == null || major < 0 || minor < 0 || patch < 0) {
            throw RynBridgeError(
                code = ErrorCode.VERSION_MISMATCH,
                message = "Invalid version format: '$version'"
            )
        }
        return SemVer(major, minor, patch)
    }

    fun isCompatible(local: String, remote: String): Boolean {
        val localVer = try { parse(local) } catch (_: Exception) { return false }
        val remoteVer = try { parse(remote) } catch (_: Exception) { return false }

        // For 0.x: minor versions must match
        if (localVer.major == 0 || remoteVer.major == 0) {
            return localVer.major == remoteVer.major && localVer.minor == remoteVer.minor
        }

        // For 1+: major versions must match
        return localVer.major == remoteVer.major
    }

    fun assertCompatible(local: String, remote: String) {
        if (!isCompatible(local, remote)) {
            throw RynBridgeError(
                code = ErrorCode.VERSION_MISMATCH,
                message = "Version mismatch: local '$local' is not compatible with remote '$remote'"
            )
        }
    }
}
